pkg/util/oidcbuilder/oidcbuilder.go

```diff
@@ -0,0 +1,105 @@
+package oidcbuilder
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/bloberror"
+
+	"github.com/Azure/ARO-RP/pkg/env"
+	utilazblob "github.com/Azure/ARO-RP/pkg/util/azblob"
+)
+
+const (
+	DiscoveryDocumentKey = ".well-known/openid-configuration"
+	JWKSKey              = "openid/v1/jwks"
+	WebContainer         = "$web"
+)
+
+type OIDCBuilder struct {
+	privateKey       []byte
+	publicKey        []byte
+	blobContainerURL string
+	endpointURL      string
+	directory        string
+}
+
+func NewOIDCBuilder(env env.Interface, oidcEndpoint string, directoryName string) (*OIDCBuilder, error) {
+	privateKey, publicKey, err := CreateKeyPair()
+	if err != nil {
+		return nil, err
+	}
+
+	return &OIDCBuilder{
+		privateKey:       privateKey,
+		publicKey:        publicKey,
+		blobContainerURL: GenerateBlobContainerURL(env),
+		endpointURL:      fmt.Sprintf("%s%s", oidcEndpoint, directoryName),
+		directory:        directoryName,
+	}, nil
+}
+
+func GenerateBlobContainerURL(env env.Interface) string {
+	return fmt.Sprintf("https://%s.blob.%s/%s", env.OIDCStorageAccountName(), env.Environment().StorageEndpointSuffix, WebContainer)
+}
+
+func (b *OIDCBuilder) EnsureOIDCDocs(ctx context.Context, azBlobClient utilazblob.AZBlobClient) error {
+	// Create the OIDC configuration
+	discoveryDocument := GenerateDiscoveryDocument(b.endpointURL)
+
+	// Create the OIDC key list
+	jwks, err := BuildJSONWebKeySet(b.publicKey)
+	if err != nil {
+		return err
+	}
+
+	return populateOidcFolder(ctx, b.directory, discoveryDocument, jwks, azBlobClient)
+}
+
+func (b *OIDCBuilder) GetEndpointUrl() string {
+	return b.endpointURL
+}
+
+func (b *OIDCBuilder) GetPrivateKey() string {
+	return string(b.privateKey)
+}
+
+func (b *OIDCBuilder) GetBlobContainerURL() string {
+	return b.blobContainerURL
+}
+
+func populateOidcFolder(ctx context.Context, directory string, discoveryDocument string, jwks []byte, azBlobClient utilazblob.AZBlobClient) error {
+	err := azBlobClient.UploadBuffer(
+		ctx,
+		"",
+		DocumentKey(directory, DiscoveryDocumentKey),
+		[]byte(discoveryDocument),
+	)
```

Comment:
Yes, it's somewhat tricky.
During the initialization of azBlobClient, the URL was passed which was already at the container level. If we pass anything instead of "", it is being added at top of the URL passed during initialization.
Eg:- https://github.com/openshift/cloud-credential-operator/blob/release-4.14/pkg/cmd/provisioning/azure/create_oidc_issuer.go#L375C1-L383C3